File: mbfpp/code/ScalaExamples/src/test/scala/com/mblinn/mbfpp/functional/fb/Selector.scala
package com.mblinn.mbfpp.functional.fb
import org.scalatest.junit.JUnitRunner
import org.scalatest.FunSpec
import org.junit.runner.RunWith
import org.scalatest.matchers.ShouldMatchers
import com.mblinn.mbfpp.functional.fb.Selector._

@RunWith(classOf[JUnitRunner])
class SelectorSpec extends FunSpec with ShouldMatchers {

  describe("A Selector Builder") {
    it("should create a selector capable of extracting a key one layer deep") {
      val ds = Map('foo -> "foo")
      selector('foo)(ds) should equal(Some("foo"))
    }

    it("should create a selector capable of extracing a key several layers deep") {
      val ds = Map('foo -> Map('bar -> Map('baz -> "baz")))
      selector('foo, 'bar, 'baz)(ds) should equal(Some("baz"))
    }

    it("should create a selector that returns None if it does not match") {
      val ds = Map('foo -> Map('bar -> Map('baz -> "baz")))
      selector('foo, 'quz)(ds) should equal(None)
    }

    it("should create a selector that returns None if an intermedia part of the datastructure is not Map[Symbol, Any]") {
      val ds = Map('foo -> Vector('bar))
      selector('foo, 'bar)(ds) should equal(None)
    }

    it("should not allow an empty path") {
      evaluating { selector() } should produce[IllegalArgumentException]
    }
  }
}

File: mbfpp/code/ClojureExamples/test/mbfpp/test/functional/fb/selector.clj
(ns mbfpp.test.functional.fb.selector
  (:use [mbfpp.functional.fb.selector])
  (:use [midje.sweet]))

(fact "Selector creates functions can extract an element one level deep"
      ((selector :foo) {:foo "foo"}) => "foo")

(fact "Selector creates functions that can extract an element several levels deep"
      ((selector :foo :bar :baz) {:foo {:bar {:baz "baz"}}}) => "baz")

(fact "Selector creates functions that return nil if they don't match"
      ((selector :foo) {:bar "bar"}) => nil)